compiler/rustc_target/src/target_features.rs

```diff
@@ -16,14 +17,31 @@ pub enum Stability {
     /// This target feature is unstable; using it in `#[target_feature]` or `#[cfg(target_feature)]`
     /// requires enabling the given nightly feature.
     Unstable(Symbol),
+    /// This feature can not be set via `-Ctarget-feature` or `#[target_feature]`, it can only be set in the basic
+    /// target definition. Used in particular for features that change the floating-point ABI.
+    Forbidden,
```

Comment:
It's written that those target features cannot be **set** (`-Ctarget-feature` or `#[target_feature]`), but can they be **conditional** used (in `#[cfg]`) ?

In `llvm_util.rs` code has been added to not print them with `--print target-features`, yet the check-cfg tests output now show that `soft-float`, `fpregs` and `x87` as possible values for `#[cfg(target_feature = "...")]`, which seems inconsistent.

*If they are not supposed to be "used", the same filtering should be applied here.* https://github.com/rust-lang/rust/blob/a4601859ae3875732797873612d424976d9e3dd0/compiler/rustc_session/src/config/cfg.rs#L373-L376

Reply:
> It's written that those target features cannot be set (-Ctarget-feature or #[target_feature]), but can they be conditional used (in #[cfg]) ?

That's a good question. The `soft-float` feature actually does differ between soft-float and hard-float targets, so I guess it makes sense to have it available in `cfg`?

I guess the question is whether `--print target-features` should print what can be set with `-Ctarget-feature`, or print what can be queried with `cfg`.

Reply:
Being able to cfg on various features seems essential, to be able to condirionally use different algorithms (e.g. fixed point vs float) or even to determine if certain compiler intrinsics should be provided to implement soft float or if they are unneeded (especially on embedded).

(I have no opinion on the output of print, if it is made for human consumption rather than machine readable you could perhaps add a `(read only, determined by target tuple)` after the entry?)

Reply:
I'm also fine with not exposing them in `cfg` for now. That's the more conservative choice as it preserves the status quo.

Reply:
So, right now the logic in `target_features` that computes which `cfg(target-feature)` are set skips forbidden features, meaning they will never be set. However, cfg-check lists *all* known features, and thus doesn't warn about asking for these target features.

We will in the future have target features that are forbidden on some targets but can unstably be set on other targets. ("forbidden" doesn't mean they are never set, it means they cannot be changed -- but the target spec can still set them, so it could still make sense to query them.) One day we will likely have features that are stable on some targets but forbidden on others. It is not clear to me how this should be handled in cfg / cfg-check -- but what this PR does for now is change nothing about cfg, and cfg-check is just a lint so we can maybe figure out its non-trivial interactions with target features later.

Reply:
I filed an issue for this: https://github.com/rust-lang/rust/issues/132351